File: lib/clusterbuster/pod_files/fio.pl
#!/usr/bin/perl

use POSIX;
use strict;
use File::Path qw(make_path remove_tree);
use JSON;
my ($dir) = $ENV{'BAK_CONFIGMAP'};
require "$dir/clientlib.pl";

my ($processes, $rundir, $runtime, $jobfiles_dir, $drop_cache_service, $drop_cache_port, $fio_blocksizes, $fio_patterns,
    $fio_iodepths, $fio_fdatasyncs, $fio_directs, $fio_ioengines, $fio_generic_args) = parse_command_line(@ARGV);

$SIG{TERM} = sub() { removeRundir() };

initialize_timing();

my ($localrundir);

sub removeRundir() {
    if (-d "$localrundir") {
	timestamp("Cleaning up run directory $localrundir");
	system('rm', '-rf', $localrundir);
    }
}

sub prepare_data_file($) {
    my ($jobfile) = @_;
    my ($filename);
    my ($filesize);
    open (my $job, '<', $jobfile) || die "Can't open job file: $!\n";
    while (<$job>) {
	chomp;
	if (/^\s*filename\s*=\s*(.+)/) {
	    $filename = $1;
	} elsif (/^\s*size\s*=\s*([[:digit:]]+)/) {
	    $filesize = $1;
	}
    }
    close $job;
    timestamp("  Job file $filename will be size $filesize");
    my ($blocksize) = 1048576;
    my ($blocks) = int(($filesize + 1048575) / $blocksize);
    my ($dirname) = $filename;
    $dirname =~ s,/[^/]*$,,;
    if (! -d $dirname) {
	system('mkdir', '-p', $dirname) || die "Can't create work directory $dirname: $!\n";
    }
    timestamp("Starting file creation");
    system('dd', 'if=/dev/zero', "of=$filename", 'bs=1048576', "count=$blocks");
    system("sync");
    timestamp("File created");
}

sub runone(;$) {
    my ($jobfile) = @_;
    my ($firsttime) = 1;
    my ($avgcpu) = 0;
    my ($weight) = .25;
    my ($icputime);
    my ($interval) = 5;
    my ($data_start_time) = xtime();
    my (%all_results);
    my ($data_start_time);

    my (@sizes) = split(/ +/, $fio_blocksizes);
    my (@patterns) = split(/ +/, $fio_patterns);
    my (@iodepths) = split(/ +/, $fio_iodepths);
    my (@fdatasyncs) = split(/ +/, $fio_fdatasyncs);
    my (@directs) = split(/ +/, $fio_directs);
    my (@ioengines) = split(/ +/, $fio_ioengines);
    my ($ucpu0, $scpu0) = cputime();
    my ($jobidx) = 1;
    my ($elapsed_time) = 0;
    timestamp("Sizes:       " . join(" ", @sizes));
    timestamp("Patterns:    " . join(" ", @patterns));
    timestamp("I/O depths:  " . join(" ", @iodepths));
    timestamp("Fdatasync:   " . join(" ", @fdatasyncs));
    timestamp("Direct I/O:  " . join(" ", @directs));
    timestamp("I/O engines: " . join(" ", @ioengines));
    timestamp("Creating workfile");
    prepare_data_file($jobfile);
    timestamp("Created workfile");
    foreach my $size (@sizes) {
	foreach my $pattern (@patterns) {
	    foreach my $iodepth (@iodepths) {
		foreach my $fdatasync (@fdatasyncs) {
		    foreach my $direct (@directs) {
			foreach my $ioengine (@ioengines) {
			    my ($jobname) = sprintf("%04d-%s-%d-%d-%d-%d-%s", $jobidx, $pattern, $size, $iodepth, $fdatasync, $direct, $ioengine);
			    drop_cache($drop_cache_service, $drop_cache_port);
			    sync_to_controller($$, $jobname);
			    if ($jobidx == 1) {
				timestamp("Running...");
				$data_start_time = xtime();
			    }
			    my ($answer0) = '';
			    timestamp("fio --rw=$pattern --runtime=$runtime --bs=$size --iodepth=$iodepth --fdatasync=$fdatasync --direct=$direct --ioengine=$ioengine $fio_generic_args --output-format=json+ $jobfile");
			    my ($jtime0) = xtime();
			    my ($jucpu0, $jscpu0) = cputime();
			    open(RUN, "-|", "fio --rw=$pattern --runtime=$runtime --bs=$size --iodepth=$iodepth --fdatasync=$fdatasync --direct=$direct --ioengine=$ioengine $fio_generic_args --output-format=json+ $jobfile | jq -c .") || die "Can't run fio: $!\n";
			    while (<RUN>) {
				timestamp($_);
				$answer0 .= "$_";
			    }
			    if (!close(RUN)) {
				timestamp("fio failed: $! $?");
				exit(1);
			    }
			    timestamp("Done job $jobfile $jobname");
			    my ($jtime1) = xtime();
			    my ($jucpu1, $jscpu1) = cputime();
			    my ($result) = from_json($answer0);
			    $jtime1 -= $jtime0;
			    $jucpu1 -= $jucpu0;
			    $jscpu1 -= $jscpu0;
			    $elapsed_time += $jtime1;
			    my (%job_result) = (
				'job_elapsed_time' => $jtime1,
				'job_user_cpu_time' => $jucpu1,
				'job_system_cpu_time' => $jscpu1,
				'job_cpu_time' => $jscpu1 + $jucpu1,
				'job_results' => $result
				);
			    $all_results{$jobname} = \%job_result;
			    $jobidx++;
			}
		    }
		}
	    }
	}
    }
    my ($data_end_time) = xtime();
    my ($ucpu1, $scpu1) = cputime();
    $ucpu1 -= $ucpu0;
    $scpu1 -= $scpu0;
    my (%extras) = (
	'results' => \%all_results
	);
    if (! ($jobfile =~ /-IGNORE-/)) {
	report_results($data_start_time, $data_end_time, $elapsed_time, $ucpu1, $scpu1, \%extras);
    }
}

sub get_jobfiles($$$) {
    my ($dir, $tmpdir, $localid) = @_;
    opendir DIR, $dir || die "Can't find job files in $dir: #!\n";

    my @files = map { "$dir/$_" } grep { -f "$dir/$_" } sort readdir DIR;
    closedir DIR;
    my (@nfiles) = ();
    foreach my $file (@files) {
	my ($nfile) = $file;
	if ($nfile =~ m,^(.*/)([^/]+)$,) {
	    $nfile = "${tmpdir}/$2";
	    timestamp("$file $1 $2 $nfile\n");
	    open READ_JOB, "<", "$file" || die "Can't open jobfile $file to read: $!\n";
	    open WRITE_JOB, ">", "$nfile" || die "Can't open temporary jobfile $file to write: $!\n";
	    while (<READ_JOB>) {
		chomp;
		timestamp($_);
		if (/^(\s*filename\s*=\s*)/) {
		    $_ .= "/$localid/$localid";
		}
		print WRITE_JOB "$_\n";
	    }
	    close READ_JOB;
	    close WRITE_JOB || die "Can't close temporary jobfile $file: $!\n";
	    push @nfiles, $nfile;
	}
    }
    print STDERR "get_jobfiles($dir) => @nfiles\n";
    return @nfiles;
}

sub runit() {
    my ($localid) = idname('-s', $$);
    $localrundir = "$rundir/$localid";
    my ($tmp_jobfilesdir) = "/tmp/fio-${localid}.job";
    mkdir "$tmp_jobfilesdir" || die "Can't create job directory $tmp_jobfilesdir: $!\n";

    removeRundir();

    if (! make_path($localrundir)) {
	timestamp("Cannot create run directory $localrundir: $!");
    }
    if (! chdir($localrundir)) {
	timestamp("Cannot cd $localrundir: $!");
	exit(1);
    }
    my (@jobfiles) = get_jobfiles($jobfiles_dir, $tmp_jobfilesdir, $localid);
    if ($#jobfiles >= 0) {
	foreach my $file (@jobfiles) {
	    runone($file);
	}
    } else {
        runone();
    }
    removeRundir();
}

run_workload($processes, \&runit);

File: lib/clusterbuster/pod_files/fio.py
#!/usr/bin/env python3

import os
import subprocess
import re
import json
import tempfile

from clusterbuster_pod_client import clusterbuster_pod_client, ClusterBusterPodClientException


class fio_client(clusterbuster_pod_client):
    """
    fio test for clusterbuster
    """

    def __init__(self):
        try:
            super().__init__()
            self._set_processes(int(self._args[0]))
            self.rundir = self._args[1]
            self.runtime = int(self._args[2])
            self.jobfilesdir = self._args[3]
            self.fio_blocksizes = self._toSizes(self._args[4])
            self.fio_patterns = re.split(r'\s+', self._args[5])
            self.fio_iodepths = self._toSizes(self._args[6])
            self.fio_numjobs = self._toSizes(self._args[7])
            self.fio_fdatasyncs = self._toBools(self._args[8])
            self.fio_directs = self._toBools(self._args[9])
            if self._args[10]:
                self.fio_ioengines = re.split(r'\s+', self._args[10])
            else:
                self.fio_ioengines = []
            self.fio_ramptime = self._toSizes(self._args[11])
            self.fio_drop_cache = self._toBool(self._args[12])
            if self._args[13]:
                self.fio_generic_args = re.split(r'\s+', self._args[13])
            else:
                self.fio_generic_args = []
        except Exception as err:
            self._abort(f"Init failed! {err} {' '.join(self._args)}")

    def get_token(self, pattern: str, string: str, token: int = 1):
        match = re.match(pattern, string)
        if match:
            return match.group(token)
        else:
            return None

    def prepare_data_file(self, jobfile: str):
        self._timestamp("Jobfile:")
        with open(jobfile) as job:
            self._timestamp(job.read())
        with open(jobfile) as job:
            lines = [line.strip() for line in job.readlines()]
            for line in lines:
                if self.get_token(r'filename\s*=\s*(.+)', line):
                    filename = self.get_token(r'\s*filename\s*=\s*(.+)', line)
                elif self.get_token(r'size\s*=\s*([0-9]+)', line):
                    filesize = int(self.get_token(r'size\s*=\s*([0-9]+)', line))
        self._timestamp(f"  Job file {filename} will be size {filesize}")
        blocksize = 2 ** 20
        blocks = int((filesize + blocksize - 1) / blocksize)
        dirname = os.path.dirname(filename)
        if not self._isdir(dirname):
            os.makedirs(dirname, exist_ok=True)
        self._timestamp("Starting file creation")
        subproc = subprocess.run(['dd', 'if=/dev/zero', f'of={filename}', f'bs={str(blocksize)}', f'count={str(blocks)}'],
                                 capture_output=True)
        if subproc.returncode != 0:
            failure = f'''
Create file failed ({subproc.returncode})
stdout: {subproc.stdout.decode()}
stderr: {subproc.stderr.decode()}
'''
            raise ClusterBusterPodClientException(failure)
        self._timestamp("File created")

    def runone(self, jobfile: str):
        elapsed_time = 0
        all_results = {}
        data_start_time = self._adjusted_time()
        ucpu, scpu = self._cputimes()
        jobidx = 1
        self._timestamp(f"""
Sizes:       {self.fio_blocksizes}
Patterns:    {self.fio_patterns}
I/O depths:  {self.fio_iodepths}
Numjobs:     {self.fio_numjobs}
Fdatasync:   {self.fio_fdatasyncs}
Direct I/O:  {self.fio_directs}
I/O engines: {self.fio_ioengines}
Drop cache:  {self.fio_drop_cache}""")
        self._timestamp("Creating workfile")
        self.prepare_data_file(jobfile)
        self._timestamp("Created workfile")
        for size in self.fio_blocksizes:
            for pattern in self.fio_patterns:
                for iodepth in self.fio_iodepths:
                    for numjobs in self.fio_numjobs:
                        for fdatasync in self.fio_fdatasyncs:
                            for direct in self.fio_directs:
                                for ioengine in self.fio_ioengines:
                                    jobname = '%04d-%s-%d-%d-%d-%d-%d-%s' % (jobidx, pattern, size, iodepth, numjobs, fdatasync, direct, ioengine)
                                    if self.fio_drop_cache:
                                        self._drop_cache()
                                    self._sync_to_controller(jobname)
                                    if jobidx == 1:
                                        self._timestamp("Running...")
                                        data_start_time = self._adjusted_time()
                                    jtime = self._adjusted_time()
                                    jucpu, jscpu = self._cputimes()
                                    with tempfile.NamedTemporaryFile() as output:
                                        outfile = output.name
                                        command = ['fio', f'--rw={pattern}', f'--runtime={self.runtime}', f'--bs={size}',
                                                   f'--iodepth={iodepth}', f'--fdatasync={int(fdatasync)}', f'--direct={int(direct)}',
                                                   f'--ioengine={ioengine}', '--allow_file_create=0', f'--numjobs={numjobs}']
                                        if not self.fio_drop_cache:
                                            command.append('--invalidate=0')
                                        command.extend(self.fio_generic_args)
                                        command.extend(['--output-format=json+', f'--output={outfile}', jobfile])
                                        success, data, stderr = self._run_command(command)
                                        if not success:
                                            err = stderr if stderr != "" else "Unknown error"
                                            raise ClusterBusterPodClientException(f'{" ".join(command)} failed: {err}')
                                        try:
                                            with open(outfile, mode='r') as f:
                                                data = f.read()
                                            result = json.loads(data)
                                        except Exception as exc:
                                            raise ClusterBusterPodClientException(f"Failed to load data {data}: {exc}")
                                    jtime = self._adjusted_time(jtime)
                                    jucpu, jscpu = self._cputimes(jucpu, jscpu)
                                    elapsed_time += jtime
                                    job_result = {
                                        'job_elapsed_time': jtime,
                                        'job_user_cpu_time': jucpu,
                                        'job_system_cpu_time': jscpu,
                                        'job_cpu_time': jucpu + jscpu,
                                        'job_results': result
                                        }
                                    all_results[jobname] = job_result
                                    jobidx = jobidx + 1
        if '-IGNORE-' not in jobfile:
            data_end_time = self._adjusted_time()
            ucpu, scpu = self._cputimes(ucpu, scpu)
            extras = {
                'results': all_results
                }
            self._report_results(data_start_time, data_end_time, elapsed_time, ucpu, scpu, extras)

    def get_jobfiles(self, jobfiledir: str, tmpdir: str, localid: str):
        new_jobfiles = []
        with os.scandir(jobfiledir) as jobfiles:
            for jobfile in jobfiles:
                jobfile = f"{jobfiledir}/{jobfile.name}"
                if not self._isfile(jobfile):
                    continue
                match = re.match(r'(.*/)([^/]+)$', jobfile)
                if match:
                    nfile = f"{tmpdir}/{match.group(2)}"
                    with open(jobfile, mode='r') as infile:
                        lines = [line.strip() for line in infile.readlines()]
                        with open(nfile, mode='wt') as outfile:
                            for line in lines:
                                if re.match(r'filename\s*=\s*', line):
                                    print(f'{line}/{localid}/{localid}', file=outfile)
                                else:
                                    print(line, file=outfile)
                    new_jobfiles.append(nfile)
        self._timestamp(f"get_jobfiles({jobfiledir}) => {' '.join(new_jobfiles)}")
        return new_jobfiles

    def runit(self, process: int):
        dirs_to_remove = []
        try:
            odir = os.getcwd()
        except Exception:
            odir = '/'
        try:
            self._cleanup_tree(self.rundir, process == 0)
            localid = self._idname(separator='-')
            localrundir = os.path.join(self.rundir, localid)
            tmp_jobsfiledir = os.path.join(self.rundir, f'fio-{localid}.job')
            self._timestamp(f"Trying to create jobdir {tmp_jobsfiledir}")
            try:
                dirs_to_remove.append(tmp_jobsfiledir)
                os.makedirs(tmp_jobsfiledir)
            except FileExistsError:
                self._timestamp(f"{tmp_jobsfiledir} already exists!")
            self._timestamp(f"Trying to create rundir {tmp_jobsfiledir}")
            try:
                dirs_to_remove.append(localrundir)
                os.makedirs(localrundir)
            except FileExistsError:
                self._timestamp(f"{tmp_jobsfiledir} already exists!")
            os.chdir(localrundir)
            jobfiles = self.get_jobfiles(self.jobfilesdir, tmp_jobsfiledir, localid)
            if not jobfiles:
                raise ClusterBusterPodClientException("Error: no jobfiles provided!")
            for jobfile in jobfiles:
                self.runone(jobfile)
        finally:
            os.chdir(odir)
            self._cleanup_tree(self.rundir, process == 0)


fio_client().run_workload()
